File: Week-1/ValidAnagram-242/solution.cpp
// Runtime: 8 ms, faster than 97.89% of C++ online submissions for Valid Anagram.
// Memory Usage: 9.4 MB, less than 88.06% of C++ online submissions for Valid Anagram.

class Solution
{
public:
    bool isAnagram(string s, string t)
    {
        if (s.size() != t.size())
            return false; //if the two strings are not the same length they cannot possibly be an anagram of each other

        int letters[26] = {0}; //create an array of size 26 to hold all letters of the alphabet

        for (int i = 0; i < s.size(); i++)
        {
            letters[s[i] - 'a']++; //for each letter in s we increment the value of letters by 1
            letters[t[i] - 'a']--; //since its an anagram each letter that occurs in s must also be in t, so to check this we
            //decrement letters by 1 for each letter than occurs in both s and t
        }

        for (int i = 0; i < 26; i++)
        {
            if (letters[i] != 0)
                return false; //if there is a element of letters that is > 0 that means there was a letter unique to one of the strings and
            //so it cannot possibly be an anagram
        }

        return true;
    }
};

File: Week-1/ValidAnagram-242/solution.js
// Runtime: 76 ms, faster than 55.92% of JavaScript online submissions for Valid Anagram.
// Memory Usage: 35.7 MB, less than 93.88% of JavaScript online submissions for Valid Anagram.

var isAnagram = function(s, t) {
  if (s.length != t.length) return false;

  var letters = new Array(26).fill(0);

  for (var i = 0; i < s.length; i++) {
    letters[s[i].charCodeAt(0) - "a".charCodeAt(0)]++;
    letters[t[i].charCodeAt(0) - "a".charCodeAt(0)]--;
  }

  for (var el in letters) {
    if (letters[el] != 0) return false;
  }

  return true;
};
